### DISCUSSION TITLE
appendix/appendix-a-trap

### DISCUSSION THREAD

**Discussion by giscus** [chai2010/advanced-go-programming-book#594] [General]
# appendix/appendix-a-trap

Go语言高级编程

https://chai2010.cn/advanced-go-programming-book/appendix/appendix-a-trap.html

**Comment by shijiehehai** [comment]
不同Goroutine之间不满足顺序一致性内存模型

**Reply by shijiehehai** [reply]
不同Goroutine之间不满足顺序一致性内存模型 这个例子 var msg string
var done bool

func setup() {
    msg = "hello, world"
    done = true
}

func main() {
    go setup()
    for !done {
    }
    println(msg)
}

**Reply by shijiehehai** [reply]
这个例子 我的go版本是1.16.6  是可以稳定输出"hello, world"的

**Reply by ByteByteDance** [reply]
多核CPU，可能出现协程a更改的变量，比如msg，在更新到内存前，就被b协程读到了

**Reply by lzwtty** [reply]
在多核处理器的计算机上，main Goroutine 和 setup Goroutine 可能在不同的处理器核心上并发执行。每个处理器核心都有自己的缓存，当执行程序时，处理器可能会将一些变量缓存在处理器的寄存器或本地缓存中，而不是立即将它们写回主内存。因此，不同 Goroutine 之间可能会看到不一致的内存状态。
在这个例子中，main Goroutine 可能在 setup Goroutine 将 msg 设置为 "hello, world" 之前，就观察到 done 变量已经变为 true。这可能是因为 done 变量的更新已经写回主内存，但 msg 变量的更新仍然缓存在 setup Goroutine 所在的处理器核心的寄存器或本地缓存中。因此，main Goroutine 看到的 msg 变量仍然是空字符串或未初始化的值。--来自gtp4
